chalice/deploy/packager.py

```diff
@@ -208,3 +205,258 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+                valid_whls.append(whl)
+            else:
+                invalid_whls.append(whl)
+
+        return valid_whls
+
+    def _install_whls(self, src_dir, dst_dir, whls):
+        # type: (str, str, List[Package]) -> None
+        if os.path.isdir(dst_dir):
+            shutil.rmtree(dst_dir)
+        os.makedirs(dst_dir)
+        for whl in whls:
+            zipfile_path = os.path.join(src_dir, whl.filename)
+            with zipfile.ZipFile(zipfile_path, 'r') as z:
+                z.extractall(dst_dir)
+
+    def build_site_packages(self, project_dir):
+        # type: (str) -> str
+        requirements_file = os.path.join(project_dir, 'requirements.txt')
+        deps_dir = self.site_package_dir(project_dir)
+        print(deps_dir)
+        if self._has_at_least_one_package(requirements_file):
+            with self._tempdir() as tempdir:
+                tempdir = os.path.abspath('whls')
+                whls = self._download_dependencies(tempdir, requirements_file)
+                self._install_whls(tempdir, deps_dir, whls)
+        return deps_dir
+
+    def site_package_dir(self, project_dir):
+        # type: (str) -> str
+        """Returns the path to the site packages directory."""
+        deps_dir = os.path.join(project_dir, '.chalice', 'site-packages')
+        return deps_dir
+
+
+class Package(object):
+    PYPI_SDIST_EXTS = ['.zip', '.tar.gz']
+
+    def __init__(self, filename):
+        # type: (str) -> None
+        self.dist_type = 'whl' if filename.endswith('whl') else 'sdist'
+        self.filename = filename
+        self.identifier = self._calculate_identifier()
+
+    def __str__(self):
+        # type: () -> str
+        return '%s(%s)' % (self.identifier, self.dist_type)
+
+    def __repr__(self):
+        # type: () -> str
+        return str(self)
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+        return self.identifier == other.identifier
+
+    def __hash__(self):
+        # type: () -> int
+        return hash(self.identifier)
+
+    def _calculate_identifier(self):
+        # type: () -> str
+        # Identiifer that can be fed into pip ie "name==version"
+        if self.dist_type == 'whl':
+            # From the wheel spec (PEP 427)
+            # {distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-
+            # {platform tag}.whl
+            name, version = self.filename.split('-')[:2]
+        else:
+            name, version = self._parse_sdist_filename(self.filename)
+        return '%s==%s' % (name, version)
+
+    def _parse_sdist_filename(self, filename):
+        # type: (str) -> Tuple[str, str]
+        # Filenames are in the format {name}-{version}.tar.gz OR
+        # {name}-{version}.zip. These are the only two formats supported by
+        # PyPi for source distributions.
+        for ext in self.PYPI_SDIST_EXTS:
+            if filename.endswith(ext):
+                name, version = filename[:-len(ext)].split('-')
```

Comment:
This is going to make you sad, but sdist filenames cannot be deterministically parsed to a single name/version without prior knowledge of what name you're looking for. For example: ``foo-1-1.tar.gz`` could be either ``("foo", "1-1")`` or ``("foo-1", "1")``.

You can unpack the tarball and ask it though, doing somehting as simple as ``python setup.py --name`` and ``python setup.py --version`` should print the name/version to stdout... although that can fail if someone writes their ``setup.py`` so that it also prints to stdout. You can do it more robustly, but uh, it's really gross to do that. You get to monkeypatch the ``setup.py`` like pip does and then run ``python setup.py egg_info`` and parse the egg-info file. I can point you to code in pip that does that.